### PR TITLE
Speedup InitialSessionState..cctor()

### PR DESCRIPTION
**Pull request by iSazonov**
# PR Summary

At startup we create a temp file with name based on `Datetime.Now.ToString()`.
It is very slow - ~14ms.
We have no need to use human date format for the temp file.

The fix is to use ~~Datetime.UtcNow().Ticks.ToString()~~ `Environment.TickCount64`:
- `Environment.TickCount64` is significantly faster Datetime.Now() - it is intrinsic and no need to evaluate _local_ time.
- Ticks is long type and Environment.TickCount64.ToString() is very fast. Datetime.ToString() is slow because it involves Calendar.

Perf win is ~10-30 ms.

Before
image

After
image

## PR Context

Tracking issue #14268

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [ ] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

File: src/System.Management.Automation/security/wldpNativeMethods.cs [hunk 1]
```diff
@@ -230,7 +230,7 @@ private static SystemEnforcementMode GetAppLockerPolicy(string path, SafeHandle
 
                                 // AppLocker fails when you try to check a policy on a file
                                 // with no content. So create a scratch file and test on that.
-                                string dtAppLockerTestFileContents = AppLockerTestFileContents + DateTime.Now;
+                                string dtAppLockerTestFileContents = AppLockerTestFileContents + DateTime.UtcNow.Ticks;
                                 IO.File.WriteAllText(testPathScript, dtAppLockerTestFileContents);
```
Comment (xtqqczze) [inline src/System.Management.Automation/security/wldpNativeMethods.cs:234]: It seems the file name does not depend on the DateTime, but rather the file **contents**. So maybe we want to keep human date format

File: src/System.Management.Automation/security/wldpNativeMethods.cs [hunk 2]
```diff
@@ -230,7 +230,7 @@ private static SystemEnforcementMode GetAppLockerPolicy(string path, SafeHandle
 
                                 // AppLocker fails when you try to check a policy on a file
                                 // with no content. So create a scratch file and test on that.
-                                string dtAppLockerTestFileContents = AppLockerTestFileContents + DateTime.Now;
+                                string dtAppLockerTestFileContents = AppLockerTestFileContents + DateTime.UtcNow.Ticks;
```
Comment (xtqqczze) [inline src/System.Management.Automation/security/wldpNativeMethods.cs:233]:
```suggestion
                                string dtAppLockerTestFileContents = AppLockerTestFileContents + DateTime.UtcNow.ToString("u");
```

File: src/System.Management.Automation/security/wldpNativeMethods.cs (hunk 1 above)
Comment (iSazonov) [reply src/System.Management.Automation/security/wldpNativeMethods.cs:234]: The file is created for _automatic_ testing - no humans see and read the file.

Comment (xtqqczze) [reply src/System.Management.Automation/security/wldpNativeMethods.cs:234]: > no humans see and read the file.

Perhaps it is for debugging purposes.

Actually, if no humans see the file do we require the datetime to be written in any form at all?

Comment (iSazonov) [reply src/System.Management.Automation/security/wldpNativeMethods.cs:234]: I don't know. I am only resolving the perf issue. I can only guess it is how AppLocker works. Perhaps @TravisEz13 or @PaulHigin could add more information.

**Comment by xtqqczze** [comment]
> It is very slow - ~14ms.

Is this a release build?

Running a microbenchmark, `Datetime.Now.ToString()` takes ~1us!

**Comment by iSazonov** [comment]
> > It is very slow - ~14ms.
> 
> Is this a release build?
> 
> Running a microbenchmark, `Datetime.Now.ToString()` takes ~1us!

Yes, it is release build. And it is startup scenario - **cold** start (.Net runtime loads Calendar data from OS and initializes internal structures - you can see this in attached trace file.). If you run BenchmarkDotNet it is **warm** start.

File: src/System.Management.Automation/security/wldpNativeMethods.cs (hunk 2 above)
Comment (xtqqczze) [reply src/System.Management.Automation/security/wldpNativeMethods.cs:233]: I took real world measurements using PerfView 😋

It seems my suggested change is not much of an improvement, I measured 8ms for the code line.

I also measured the original code line and it was 14ms, same as your measurement.

Comment (iSazonov) [reply src/System.Management.Automation/security/wldpNativeMethods.cs:233]: DateTime.UtcNow is more fast - it does not read current user time zone and does not convert to local time.

File: src/System.Management.Automation/security/wldpNativeMethods.cs (hunk 1 above)
Comment (daxian-dbw) [reply src/System.Management.Automation/security/wldpNativeMethods.cs:234]: @TravisEz13 and @PaulHigin, can you please comment on whether we need `DateTime` to be added to the file content at all?

Comment (PaulHigin) [reply src/System.Management.Automation/security/wldpNativeMethods.cs:234]: The DateTime was included to add some randomness to the file contents.  I think we still need it, but could change it to some other random value if computing it is quicker than DateTime.

Comment (daxian-dbw) [reply src/System.Management.Automation/security/wldpNativeMethods.cs:234]: @iSazonov Maybe replacing `DateTime.Now.Ticks` with `Environemnt.TickCount`?

Comment (iSazonov) [reply src/System.Management.Automation/security/wldpNativeMethods.cs:234]: @daxian-dbw I have no objections. I see `Environment.TickCount64` - maybe use it?

Comment (daxian-dbw) [reply src/System.Management.Automation/security/wldpNativeMethods.cs:234]: That sounds fine to me, go ahead.

Comment (iSazonov) [reply src/System.Management.Automation/security/wldpNativeMethods.cs:234]: Done. The PR description has been updated too.

**Review by daxian-dbw — APPROVED** [review]

**Pull request merged**